src/gdb_packet.c

```diff
@@ -31,119 +31,220 @@
+				/* 
+				 * Let consume_remote_packet handle this
+				 * returns PACKET_IDLE or PACKET_GDB_CAPTURE if it detects the start of a GDB packet
 				 */
-				packet[0] = '\0';
+				state = consume_remote_packet(packet, size);
+				/* No need to handle checksum or offset, everything happened in PACKET_IDLE state so nothing was modified */
 			}
 #endif
-		} while (packet[0] != '$');
-
-		offset = 0;
-		csum = 0;
-		char c = '\0';
-		/* Capture packet data into buffer */
-		while (c != '#') {
-			c = gdb_if_getchar();
-			if (c == '#')
-				break;
-			/* If we run out of buffer space, exit early */
-			if (offset == size)
-				break;
+			break;
 
-			if (c == '$') { /* Restart capture */
+		case PACKET_GDB_CAPTURE:
+			if (rx_char == GDB_PACKET_START) {
+				/* Restart GDB packet capture */
 				offset = 0;
-				csum = 0;
-				continue;
+				checksum = 0;
+				break;
 			}
-			if (c == '}') { /* Escaped char */
-				c = gdb_if_getchar();
-				csum += c + '}';
-				packet[offset++] = (char)((uint8_t)c ^ 0x20U);
-				continue;
+			if (rx_char == GDB_PACKET_END) {
+				/* End of GDB packet */
+				// if (noackmode)
+				// 	/* knowlingly ignore the checksum that follows in no ack mode (as per spec) */
+				// 	state = PACKET_GDB_COMPLETE;
+				// else
+				/* Move to checksum capture */
+				state = PACKET_GDB_CHECKSUM_UPPER;
+				break;
 			}
-			csum += c;
-			packet[offset++] = c;
-		}
-		recv_csum[0] = gdb_if_getchar();
-		recv_csum[1] = gdb_if_getchar();
-		recv_csum[2] = 0;
 
-		/* Return packet if checksum matches */
-		if (csum == strtoul(recv_csum, NULL, 16))
+			/* Not start or end of packet, add to checksum */
+			checksum += rx_char;
+
+			/* Add to packet buffer, unless it is an escape char */
+			if (rx_char == GDB_PACKET_ESCAPE)
+				/* GDB Escaped char */
+				state = PACKET_GDB_ESCAPE;
+			else
+				/* Add to packet buffer */
+				packet[offset++] = rx_char;
 			break;
 
-		/* Get here if checksum fails */
-		gdb_if_putchar('-', 1); /* Send nack */
-	}
-	gdb_if_putchar('+', 1); /* Send ack */
-	packet[offset] = '\0';
+		case PACKET_GDB_ESCAPE:
+			/* Add to checksum */
+			checksum += rx_char;
 
-	DEBUG_GDB("%s: ", __func__);
-	for (size_t j = 0; j < offset; j++) {
-		const char value = packet[j];
-		if (value >= ' ' && value < '\x7f')
-			DEBUG_GDB("%c", value);
-		else
-			DEBUG_GDB("\\x%02X", value);
+			/* Resolve escaped char */
+			packet[offset++] = rx_char ^ GDB_PACKET_ESCAPE_XOR;
+
+			/* Return to normal packet capture */
+			state = PACKET_GDB_CAPTURE;
+			break;
+
+		case PACKET_GDB_CHECKSUM_UPPER:
+			/* Checksum upper nibble */
+			rx_checksum = unhex_digit(rx_char) << 4; /* This also clears the lower nibble */
+			state = PACKET_GDB_CHECKSUM_LOWER;
+			break;
+
+		case PACKET_GDB_CHECKSUM_LOWER:
+			/* Checksum lower nibble */
+			rx_checksum |= unhex_digit(rx_char); /* BITWISE OR lower nibble with upper nibble */
+
+			// 	state = PACKET_GDB_COMPLETE;
+			// 	break;
+			// case PACKET_GDB_COMPLETE:
```

Comment:
Dead code?

Reply:
This is part of a temporary commit to get things works, this is why I noted "theres still an issue" this wont be part of the final submission

Reply:
Ah, excellent. Thank you :)